contracts/src/EspToken.sol

```diff
@@ -0,0 +1,46 @@
+// contracts/GLDToken.sol
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+// TODO MA: should we really use solmate, or just stick with openzeppelin?
+import { ERC20Upgradeable } from
+    "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import { OwnableUpgradeable } from
+    "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import { Initializable } from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import { UUPSUpgradeable } from
+    "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+// TODO: Timelock
+contract EspToken is Initializable, ERC20Upgradeable, OwnableUpgradeable, UUPSUpgradeable {
+    /// @notice upgrade event when the proxy updates the implementation it's pointing to
+    event Upgrade(address implementation);
+
+    /// @notice since the constructor initializes storage on this contract we disable it
+    /// @dev storage is on the proxy contract since it calls this contract via delegatecall
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    /// @notice This contract is called by the proxy when you deploy this contract
+    function initialize(address _owner, address _initialGrantRecipient) public initializer {
+        __Ownable_init(_owner);
+        __UUPSUpgradeable_init();
+        _mint(_initialGrantRecipient, 1000_000_000 ether);
+    }
+
+    /// @notice only the owner can authorize an upgrade
+    function _authorizeUpgrade(address newImplementation) internal override onlyOwner {
```

Comment:
it could be interesting to add some validation checks here to ensure that the new implementation is valid e.g. a check that the version on the new implementation is greater than the current version

Reply:
Actually it sounds like we need to do more here and actually check msg.sender?

https://docs.openzeppelin.com/contracts/5.x/api/proxy#UUPSUpgradeable-_authorizeUpgrade-address-

Reply:
ah no it has the `onlyOwner` modifier, maybe adding a test wouldn't be a bad idea then.

Reply:
On a related note. UUPSUpgradable already emits an `Upgraded(address newImpl)` event: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/fda6b85f2c65d146b86d513a604554d15abd6679/contracts/proxy/ERC1967/ERC1967Utils.sol#L69

Reply:
I thought the purpose of this function was mainly to check if someone has the right to upgrade but not if the upgrade is "good". I think we need to test our upgrades well but I'm not sure this function can do much for us in this regard.